### PR TITLE
Add Regex span startat overloads

### LINKED ISSUE #68959: [API Proposal]: Additional Regex.EnumerateMatches/Count overloads
### Background and motivation

The existing `Regex.Matches` method has an overload that accepts an `int startat`, but the newly added `EnumerateMatches` method lacks such an overload.  While you can slice the input span to begin at some position into the input, there's a difference in regex between "beginning" and "starting".  The former means there's no available input prior to where the matching process first runs, whereas the latter means there may be available input prior to where the matching process first runs.  That's relevant for lookbehinds, as a lookbehind can see text prior to a "startat" but not prior to a "beginning".  And it means uses like:
https://github.com/dotnet/runtime/blob/ed2472b72679ff5ca0fa56996cf15d2fc303c7db/src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs#L195-L200
which could otherwise shift to using `EnumerateMatches` for a performance gain are unable to do so, as this particular case has a lookbehind and uses a potentially positive startat value.

### API Proposal

```diff
public class Regex
{
    // Existing
    public MatchCollection Matches(string input, int startat);
    public Match Match(string input, int startat);
    public bool IsMatch(string input, int startat);

    // Previously added in .NET 7
    public int Count(System.ReadOnlySpan<char> input);
    public Regex.ValueMatchEnumerator EnumerateMatches(System.ReadOnlySpan<char> input);

    // Proposed
+   public int Count(System.ReadOnlySpan<char> input, int startat);
+   public Regex.ValueMatchEnumerator EnumerateMatches(System.ReadOnlySpan<char> input, int startat);
}
```

### API Usage

```csharp
foreach (ValueMatch m in _regex.EnumerateMatches(input, startat))
{
    ...
}
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by stephentoub**
Also fixes the new Count methods to behave correctly for RightToLeft.

Fixes https://github.com/dotnet/runtime/issues/68959

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Count.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ public int Count(string input)
 
             int count = 0;
 
-            RunAllMatchesWithCallback(input, 0, ref count, static (ref int count, Match match) =>
+            RunAllMatchesWithCallback(input, RightToLeft ? input.Length : 0, ref count, static (ref int count, Match match) =>
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Count.cs:23]: Good catch! I would have expected tests to catch this, but I'm glad you now added tests for it 👍

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.EnumerateMatches.Tests.cs [hunk 2]
```diff
@@ -151,16 +151,25 @@ public partial class RegexCountTests
...
-        public void EnumerateMatches_ReturnsExpectedCount(RegexEngine engine, string pattern, string input, RegexOptions options, int expectedCount)
+        public void EnumerateMatches_ReturnsExpectedCount(RegexEngine engine, string pattern, string input, int startat, RegexOptions options, int expectedCount)
         {
             Regex r = RegexHelpers.GetRegexAsync(engine, pattern, options).GetAwaiter().GetResult();
-            int count = 0;
-            foreach (ValueMatch _ in r.EnumerateMatches(input))
+
+            int count;
+
+            count = 0;
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.EnumerateMatches.Tests.cs:160]: Just curious, why is this preferred than `int count = 0;`?

**Review by joperezr — APPROVED** [review]
LGTM. sorry for the wait, I was out Thursday and Friday.

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.EnumerateMatches.Tests.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.EnumerateMatches.Tests.cs:160]: Doesn't really matter either way.  I'd separated it when I had multiple similar blocks and wanted them all to be consistent in terms of setting count to 0, and it felt inconsistent having one of them declare the varaible.

**Pull request merged**